### ISSUE TITLE
Negative loss during contrastive learning

### ISSUE THREAD

**Issue by yunzz** [marukosan93/RS-rPPG#10]
Hi!
I encountered a problem during the contrastive learning stage.
The loss keeps going negative. After some logging and debugging, I found that the situation looks like this: the positive sample differences are much larger than the negative sample differences, which makes it seem as if the positives and negatives are swapped, leading to a naturally negative loss.

Here is a full log snippet (batch 0):
===== tinit stage input difference check (batch 0) =====
MSTmap mean: 0.042990 | Tmap mean: 0.481122 | Difference: 0.438132
Sample sorting finished!
Start training: total 50 epochs, batch size 4, learning rate 0.0001

===== Training epoch 1/50 =====

===== Input data difference check (batch 0) =====
Anchor mean: 0.075563 | Positive mean: 0.487652 | Difference: 0.412089
Freq negative mean: 0.067529 | Difference: 0.008033
Background negative mean: 0.508153 | Difference: 0.432590
Motion negative mean: 0.433993 | Difference: 0.358430

===== Model feature std check (batch 0) =====
Anchor feature std: 0.999538 | Positive: 0.999661 | Freq negative: 0.999540

===== Negative sample index check (batch 0) =====
Batch indices: [0 1 2 3] | Negative1 indices: (1, 3, 0, 2)
Negative1 same as anchor output? False (expected False)

===== PSD difference check (batch 0) =====
Anchor PSD: 0.005882 | Positive PSD: 0.005882 | Difference: 0.000000
Negative1 PSD: 0.005882 | Difference: 0.000000
Negative2 PSD: 0.005882 | Difference: 0.000000
Warning: Negatives too similar to anchor → contrastive learning fails？

Epoch: [0/49][0/457]    Time 3.532 (3.532)   Data 0.499 (0.499)   Loss -0.0270 (-0.0270)   Loss_sp 0.3818 (0.3818)   Loss_bw 0.9739 (0.9739)

From my pretraining stage, the Tmaps looked fine, so I guess the problem is not in the tmap2 generation.
What confuses me even more is that during contrastive learning the MAE/RMSE are actually worse than in the pretraining stage, which seems very strange.

I also tested different tau values (10 and 0.1), but the problem remains.

Do you have any idea what might cause this? Could it be an issue with my dataloader, or something wrong in how I construct positives/negatives?

This is my first time running such a model, so I might have overlooked something obvious.
Thanks a lot for your guidance!

**Comment by marukosan93** [comment]
Most likely an issue with the data processing, or the sampling, could be anything really. What helps is usually trying on an easier dataset first (like PURE,OBF) and then moving on to harder ones like VIPL-HR. As there you can easier sort out the issues and diagnose the problem as the task is simpler. The A-P distances should decrease while training and the A-N distances should increase.

**Comment by yunzz** [comment]
Thanks a lot for the suggestion! I’ll try running on PURE/OBF first to verify the sampling and distances, and then move to VIPL-HR again.
